Rewrite cmd/linkcheck

caesarxuchao opened the pull request:
Fix #15916, #16233.

Now the linkcheck actually visits a url and checks the response code. It reports url as invalid if the status code >=400.

This check requires a long time to run. Checking docs/ folder alone takes around 5 minutes. Some possible improvements:
1. only run the check on the diff of a PR;
2. add a whitelist of valid URLs, and update the whitelist periodically.

Any suggestions?
cc @lavalamp @nikhiljindal @bgrant0607 @timstclair

caesarxuchao commented:
I'll make it a cron job that runs on Jenkins once a day.

caesarxuchao commented:
First bad link found by the tool:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/extensions/v1beta1/types.go#L313

caesarxuchao commented:
@nikhiljindal A friendly ping.

nikhiljindal commented:
`hack/verify-godeps.sh` is failing with `Your godeps changes are not reproducable`.
@caesarxuchao Can you please take a look?

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
-//$ linkcheck --root-dir=${TYPEROOT} --repo-root=${KUBE_ROOT} \
-//  --file-suffix=types.go --prefix=http://releases.k8s.io/HEAD
+// This tool extracts the links from types.go and .md files, visits the link and
+// check the status code of the response.
```
check -> checks

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
-	fileSuffix = flag.String("file-suffix", "", "suffix of files to be checked")
-	prefix     = flag.String("prefix", "", "Longest common prefix of the link URL, e.g., http://release.k8s.io/HEAD/ for links in pkg/api/types.go")
+	fileSuffix = flag.StringSlice("file-suffix", []string{"types.go", ".md"}, "suffix of files to be checked")
+	whiteList  = []*regexp.Regexp{
```
Add a comment saying what is means to be in this whitelist?
When should someone add a link to this list?

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
+		regexp.MustCompile(`http://kubernetes\.io/img/warning\.png`),
+		regexp.MustCompile(`http://releases\.k8s\.io$`),
+		// skip url that doesn't start with an English alphabet.
+		regexp.MustCompile(`https?://[^A-Za-z].*`),
```
Why this?
Do we have such URLs?

caesarxuchao replied:
In examples, there are url to a special host's IP address, we need to skip those.

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -14,39 +14,47 @@ See the License for the specific language governing permissions and
+func newWalkFunc(invalidLink *bool, fileSuffixTable map[string]struct{}, client *http.Client) filepath.WalkFunc {
 	return func(filePath string, info os.FileInfo, err error) error {
-		if !strings.HasSuffix(info.Name(), *fileSuffix) {
+		_, found1 := fileSuffixTable[filepath.Ext(filePath)]
```
Its not clear what are you trying to find here.
Maybe use better variable name than found1 and found2?

caesarxuchao replied:
We want to check is file is *.md or types.go, one is an extension, the other one is a file name, so we need to search twice. I'll add a comment.

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
-			//match[1] should look like docs/devel/api-conventions.md
-			if _, err := os.Stat(path.Join(*repoRoot, string(match[1]))); err != nil {
-				fmt.Fprintf(os.Stderr, "Link is not valid: %s\n", string(match[0]))
+		allURL := xurls.Strict.FindAll(fileBytes, -1)
```
allURL -> allURLs

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
+			// remove the htmlpreview Prefix
+			processedURL := htmlpreviewReg.ReplaceAll(URL, []byte{})
+			retry := 0
+			const maxRetry int = 3
```
optional: Make this a flag?

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
+					break
+				}
+				if resp.StatusCode == 429 {
+					fmt.Fprintf(os.Stderr, "Got %d visiting %s, retry after %d seconds.\n", resp.StatusCode, string(URL), backoff)
```
The right way is to read the "Retry-after" header which should tell you the number of seconds you should wait before making another call.

caesarxuchao replied:
Oh, thanks! Let me try it out.

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -54,16 +62,52 @@ func newWalkFunc(invalidLink *bool) filepath.WalkFunc {
+				// We only check response code >= 400. We don't care about
+				// errors.
+				if err != nil {
+					break
```
Should you maybe retry? Silently ignoring doesnt seem right

caesarxuchao replied:
Some error I noted down, they are cuased by fake URLs:
```
	// Unexpected error: Get http://server:port: dial tcp: unknown port tcp/port
	// Unexpected error: Get http://fed-master:4001/v2/keys/coreos.com/network/subnets: dial tcp: lookup fed-master: no such host
	// Unexpected error: Get http://kube-master:8080/api/v1: dial tcp: lookup kube-master: no such host
	// Unexpected error: Get http://gitserver.$(POD_NAMESPACE):$(SERVICE_PORT): dial tcp: unknown port tcp/$(SERVICE_PORT)
	// Unexpected error: Get http://ip:port: dial tcp: unknown port tcp/port
```

I'll check again tomorrow, if the number of such links is small, I'll just add them to the whiteList.

nikhiljindal commented on hack/after-build/verify-linkcheck.sh:
```diff
@@ -24,15 +24,20 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
+for root in "${ROOTS[@]}"; do
+  "${linkcheck}" "--root-dir=${root}" && ret=0 || ret=$?
+  if [[ $ret -eq 1 ]]; then
+    echo "links in ${root} is out of date."
```
Shouldnt it be "found invalid links in ${root}"

caesarxuchao replied:
Good catch. Thanks.

nikhiljindal commented on hooks/pre-commit:
```diff
@@ -112,28 +112,6 @@ else
-echo "${reset}"
-
-echo -ne "Checking for docs that need updating... "
-if ! hack/after-build/verify-generated-docs.sh > /dev/null; then
```
Why removing this?

caesarxuchao replied:
A slip. I will add it back. Thanks.

nikhiljindal commented:
Thanks! Added a few comments.

cc @mvdan who might be interested in our usage of github.com/mvdan/xurls

caesarxuchao commented:
I addressed most comments and left an note on why we'd better keep ignoring the error.

One more thing we need to fix before merging: our munger generates `The latest release of this document can be found here.` no matter if the doc actually exists in the latest release, the link checker reports many such errors.

nikhiljindal commented:
@caesarxuchao Whats the status of this?

caesarxuchao commented:
It's in my todo list this week :)

caesarxuchao commented:
The output of hack/verify-linkcheck.sh, all found invalid links are really invalid:
```
Summary of invalid links:
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/pkg/apis/extensions/v1beta1/types.go, Got 404 visiting http://releases.k8s.io/HEAD/docs/admin/daemon.md
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/design/event_compression.md, Got 404 visiting http://releases.k8s.io/HEAD/pkg/client/unversioned/record/event.go
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/coreos/bare_metal_offline.md, Got 404 visiting https://storage.googleapis.com/kubernetes-release/release/v0.15.0/bin/linux/amd64/kubecfg
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/coreos/bare_metal_offline.md, Got 404 visiting https://storage.googleapis.com/k8s/flanneld
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/logging.md, Got 404 visiting http://releases.k8s.io/release-1.1/contrib/logging/fluentd-sidecar-gcp/README.md
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/rackspace.md, Got 404 visiting https://github.com/kubernetes/kubernetes/archive/v0.9.tar.gz
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/getting-started-guides/ubuntu-calico.md, Got 404 visiting https://github.com/kubernetes/kubernetes/releases/download/v1.1.0/kubernetes.tar.gz
Failed: in file /usr/local/google/home/xuchao/go-workspace/src/k8s.io/kubernetes/docs/user-guide/production-pods.md, Got 404 visiting http://releases.k8s.io/HEAD/contrib/git-sync/
```

caesarxuchao commented:
@nikhiljindal, PTAL. Thanks!

mvdan commented:
Thanks for the ping @nikhiljindal!

@caesarxuchao what do you mean by this?

    // xurls is confused by http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl.

caesarxuchao commented:
Hi @mvdan, I mean when parsing "http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl.", xurls identifies ""http://storage.googleapis.com/kubernetes-release/release/v$" as a url, rather than the whole string, so I add the partial string into the white list.

xurls is very easy to use, thanks for providing it!

caesarxuchao commented on cmd/linkcheck/links.go:
```diff
@@ -14,56 +14,158 @@ See the License for the specific language governing permissions and
+		"http://stackoverflow.com/questions/ask?tags=kubernetes":  {},
+		"https://github.com/$YOUR_GITHUB_USERNAME/kubernetes.git": {},
+		"https://github.com/$YOUR_GITHUB_USERNAME/kubernetes":     {},
+		// xurls identifies the part before "{" in URL "http://storage.googleapis.com/kubernetes-release/release/v${K8S_VERSION}/bin/darwin/amd64/kubectl" as an URL.
```
@mvdan, I reworded the comment to be more specific.

Commit:
linkchecker tool now visits the URL to determine if it's valid

Commit:
update godep

mvdan commented:
@caesarxuchao I have added support for braces in https://github.com/mvdan/xurls/commit/1b768d7c393abd8e8dda1458385a57becd4b2d4e, this should fix it for you.

caesarxuchao commented:
@mvdan, yes, that's fixed. Thank you!

nikhiljindal commented:
Added a few minor comments to https://github.com/caesarxuchao/kubernetes/commit/12c1a1910cece8690efb79781c7d7e7a5d054959.

caesarxuchao commented:
@nikhiljindal I pushed another commit to address the comments. PTAL. Thanks.

nikhiljindal commented on cmd/linkcheck/links.go:
```diff
@@ -14,56 +14,155 @@ See the License for the specific language governing permissions and
+	// efficiency.
+	regWhiteList = []*regexp.Regexp{
+		regexp.MustCompile(`https://kubernetes-site\.appspot\.com`),
+		// skip url that doesn't start with an English alphabet.
```
Say in the comment that this is for URLs with IP addresses.

caesarxuchao replied:
Done.

nikhiljindal commented on Makefile:
```diff
@@ -49,7 +49,6 @@ verify:
 	hack/verify-generated-deep-copies.sh
 	hack/verify-generated-docs.sh
 	hack/verify-swagger-spec.sh
-	hack/verify-linkcheck.sh
```
Why this change?

caesarxuchao replied:
This link tool will not be run in Jenkins unit-test or e2e builds, it will have its own Jenkins project which runs once every day, because checking all URLs takes too much time.

caesarxuchao commented:
@nikhiljindal, I addressed your comments. Could you take another look? Thanks.

nikhiljindal commented:
LGTM, thanks!

nikhiljindal commented:
Feel free to self apply the LGTM label after you have squashed the commits.

caesarxuchao commented:
Apply the LGTM per nikhil's comment.

ixdy commented:
@spxtr can you help @caesarxuchao get `kubernetes-check-links` set up again with a proper checked-in Jenkins config now that this is ready to be checked in?

spxtr commented:
Sure.

@caesarxuchao If you want this to run regularly on Jenkins you'll need to add a job config for it in `hack/jenkins/job-configs`. You can probably copy most of `kubernetes-test-go`'s config, but you won't need to use a template. See [Jenkins Job Builder](http://docs.openstack.org/infra/jenkins-job-builder/) for docs, and bother me with questions.

caesarxuchao commented:
A straight-forward rebase, adding back the label.

alex-mohr commented:
@k8s-bot test this

caesarxuchao commented:
travis failure:
```
Your godeps changes are not reproducable
diff -Naupr '--exclude=symlink' /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt /tmp/gopath.4eSANu/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt
--- /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	1970-01-01 00:00:00.000000000 +0000
+++ /tmp/gopath.4eSANu/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	2016-01-21
```

I vaguely remember there are changes to this licence verification. I'll try if rebase solve the problem.

Commit:
run update-godep-licenses.sh

caesarxuchao commented:
@karlkfi do you have any idea why `hack/verify-godep-license.sh` complains about Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt? The repo doesn't have that file and this PR doesn't touch that folder.

caesarxuchao commented:
@k8s-bot e2e test this please

caesarxuchao commented:
#19977 should have fixed the godep error.

caesarxuchao commented:
Just rebase, so adding back the lgtm.
